Implement wind estimation

fiam opened the pull request:
- [x] Import code from PR #2284
- [x] Move code into `navigation/`
- [x] Use functions from navigation system rather than checking GPS and IMU directly.
- [x] Fix code (probably a lot of errors still in there)
- [x] Implement basic OSD indicators
- [x] Full OSD indicators (symbols, proper unit support)
- [x] Decide wether we want to show the wind relative to earth with just YAW rotation or wind relative to aircraft
- [x] Reset OSD indicators in `pgResetFn_osdConfig()`


Configurator support at https://github.com/iNavFlight/inav-configurator/pull/300

Linjieqiang committed:
Use GPS and DCM to estimate the airspeed for fixwing.

fiam committed:
Move wind estimator to navigation

- Move est_wind_airspeed to navigation_wind_estimator
- Make a few stylistical changes in all the file to use
INAV conventions and constants for better readability.
- Add function for retrieving wind estimates in body frame.
- Mark with missing stuff with TODO

fiam committed:
Initial, very basic implementation of OSD wind indicators

An indicator shows the wind in the XY plane while another shows
the Z axis.

fiam committed:
Use dedicated symbols for OSD wind indicators

fiam committed:
Rename OSD field constants for wind indicators

Cosmetic change

OSD_HORIZONTAL_WIND -> OSD_WIND_SPEED_HORIZONTAL
OSD_VERTICAL_WIND -> OSD_WIND_SPEED_VERTICAL

fiam committed:
Respect OSD units when formatting wind speed

Move formatting to a new osdFormatWindSpeedStr() function
which is used for both wind speed indicators.

fiam committed:
Only estimate wind speed in fixed wing

fiam committed:
Check gpsSol flags for valid VelNE and VelD before using them

fiam committed:
Remove double ;

Thanks to @giacomo892 for pointing it

digitalentity commented on src/main/navigation/navigation_wind_estimator.c:
```diff
@@ -0,0 +1,181 @@
+        memcpy(lastFuselageDirection, fuselageDirection, sizeof(lastFuselageDirection));
+        memcpy(lastGroundVelocity, groundVelocity, sizeof(lastGroundVelocity));
+
+        float theta = atan2f(groundVelocityDiff[1], groundVelocityDiff[0]) - atan2f(groundVelocityDiff[1], groundVelocityDiff[0]);// equation 9
```
We need to use fastmath functions here: `atan2_approx`, `sin_approx` and `cos_approx`.

fiam committed:
Move check for valid GPS heading to a function

Allows calling it from both the IMU and the wind estimator

fiam committed:
Check for valid GPS heading, not valid nav system heading

We're using just information from the GPS, so having a valid heading
due to the presence of a compass won't work. This also makes this
code not depend on the NAV system anymore, so we'll move it
somewhere else.

fiam committed:
Tweak the filtering for estimated wind speed a bit

Remove conditional which allowed only negative or small positive
changes to the estimated wind speed. Leave the current 0.95 and
0.05 weights for past estimate/current estimate. Eventually we'll
add a proper filter.

fiam committed:
Remove unused include

fiam committed:
Calculate wind speed in cm/s

Due to incorrect scaling, it was calculated in m/s

fiam committed:
Add comments explaning how the wind estimator works

fiam committed:
Change wind velocity presentation to the user in OSD

Don't change speed in Z axis, always show vertical wind speed in
earth frame.
For the XY plane, just calculate magnitude and angle relative to
the aircraft in osd.c. This way we don't need to perform a full
3D rotation.

This lets use remove the getEstimatedWindVelocityBodyFrame()
function.

fiam committed:
When wind estimation isn't valid, show indicators as '---'

Allows users to tell apart unknown wind from wind detected as
zero.

fiam committed:
Make wind estimator require GPS support

It doesn't depend on NAV anymore, just GPS

fiam committed:
Move wind estimator to flight/, call it only on GPS updates

- Move navigation_wind_estimator.{c,h} to flight/wind_estimator.{c,h}
- Rename gpsThread() to gpsUpdate(), make it return a bool which is
true iff new GPS data was received.
- When the GPS returns new data, call into the wind estimator to
update itself.
- Make sure wind estimator can be disabled at compile time via
the USE_WIND_ESTIMATOR #ifdef
- Disable OSD wind indicators when wind estimator support is not
compiled in.
- Don't check for both GPS and USE_WIND_ESTIMATOR in ifdefs, instead
fail compilation with an error in USE_WIND_ESTIMATOR is defined
without defining GPS.

fiam committed:
Add Wind Estimation paper by William Premerlani

Used to implement the wind estimator

fiam committed:
Add back check for bounding wind updates

Allow either updates which lower the total wind speed or increase
it slightly. Arduplane uses a similar algorithm, so let's fly
test this and see what happens.

junwoo091400 commented on src/main/flight/wind_estimator.c:
```diff
@@ -0,0 +1,155 @@
+        memcpy(lastFuselageDirection, fuselageDirection, sizeof(lastFuselageDirection));
+        memcpy(lastGroundVelocity, groundVelocity, sizeof(lastGroundVelocity));
+
+        float theta = atan2f(groundVelocityDiff[1], groundVelocityDiff[0]) - atan2f(groundVelocityDiff[1], groundVelocityDiff[0]);// equation 9
```
Is this correct?? Wouldnt this give 0 every time??

alexbojko replied:
as far as I understand from documentation here should be the difference between vectors of ground velocity speed and fuselage direction. I think it will work:

`
float theta = atan2f(groundVelocityDiff[1], groundVelocityDiff[0]) - atan2f(fuselageDirectionDiff[1], fuselageDirectionDiff[0]);
`

junwoo091400 replied:
Yes, I didn't see the document, but your suggestion seems better.

fiam committed:
Fix incorrect calculation for theta during wind estimation

Kudos to @junwoo091400 and @alexbojko for spotting it

fiam committed:
Add WIND_ESTIMATOR debug mode

Uses debug[0-2] to log estimated wind in earth frame X, Y, Z.

fiam committed:
Fix value of DEBUG_WIND_ESTIMATOR

After latest rebase, it should be placed after DEBUG_STAGE2

fiam committed:
Fix threshold for maximum estimated wind length increase

- Threshold was in m/s, while the wind is calculated in cm/s.
- Fix debug[] values, no need to *100 since wind is already cm/s.
- Replace a runtime sqrtf() with a compile time sq().
- Rename a few variables to snakeCase.

fiam committed:
Increase wind measurement timeout to 10s

fiam committed:
Re-enable OSD wind speed indicators

shellixyz commented on src/main/io/osd.c:
```diff
@@ -352,6 +354,39 @@ static void osdFormatVelocityStr(char* buff, int32_t vel)
+            suffix = SYM_MPH;
+            break;
+        case OSD_UNIT_METRIC:
+            centivalue = (ws * 36) / 10;;
```
There are 2 semicolons at the end of this line

fiam replied:
Nice catch! Fixed :-)

shellixyz commented on src/main/io/osd.h:
```diff
@@ -81,6 +81,8 @@ typedef enum {
     OSD_MAP_TAKEOFF,
     OSD_RADAR,
     OSD_DEBUG, // Number 46. Intentionally absent from configurator and CMS. Set it from CLI.
```
Shouldn't OSD_DEBUG be last ?

fiam replied:
(we discussed this, but let's add it here for others)

OSD_DEBUG is just an OSD indicator that we don't show in CMS nor in the configurator. From the firmware PoV there's nothing special about it.

shellixyz reviewed (commented):
I'm going to test again it again today with the modified thresholds and the OSD indicators.

fiam committed:
Remove extra semicolon at end of line

fiam committed:
Refactor calculation for horizontal wind speed to wind_estimator.c

Add float getEstimatedHorizontalWindSpeed(uint16_t *angle)
